### PR TITLE
Create ticker on metachain

### PR DESCRIPTION
**Pull request by sasurobert**
Add ticker on metachain
Add interface to process built in functions easily on metachain

### CODE REVIEW

File: vm/systemSmartContracts/liquidStaking.go [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+
+	l := &liquidStaking{
+		eei:                      args.Eei,
+		delegationMgrSCAddress:   args.DelegationMgrSCAddress,
+		endOfEpochAddr:           args.EndOfEpochAddress,
+		liquidStakingSCAddress:   args.LiquidStakingSCAddress,
+		gasCost:                  args.GasCost,
+		marshalizer:              args.Marshalizer,
+		hasher:                   args.Hasher,
+		liquidStakingEnableEpoch: args.EpochConfig.EnableEpochs.BuiltInFunctionOnMetaEnableEpoch,
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/liquidStaking.go:81]: BuiltInFunctionOnMetaEnableEpoch actually represents the enable epoch for liquid staking?

File: vm/systemSmartContracts/liquidStaking.go [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+func (l *liquidStaking) Execute(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	l.mutExecution.RLock()
+	defer l.mutExecution.RUnlock()
+
+	err := CheckIfNil(args)
+	if err != nil {
+		l.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+	if !l.flagLiquidStaking.IsSet() {
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/liquidStaking.go:100]: Maybe this check could be moved before L95

File: vm/systemSmartContracts/liquidStaking.go [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+	if len(args.ESDTTransfers) < 1 {
+		l.eei.AddReturnMessage("function requires liquid staking input")
+		return vmcommon.UserError
+	}
+	if args.CallValue.Cmp(zero) != 0 {
+		l.eei.AddReturnMessage("function is not payable in eGLD")
+		return vmcommon.UserError
+	}
+	for _, esdtTransfer := range args.ESDTTransfers {
+		if !bytes.Equal(esdtTransfer.ESDTTokenName, l.getTokenID()) {
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/liquidStaking.go:157]: tokenID := l.getTokenID() outside this for and then you can use directly tokenID here, instead calling l.getTokenID() at each iteration

File: vm/systemSmartContracts/liquidStaking.go [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+	return vmcommon.UserError
+}
+
+func (l *liquidStaking) init(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if !bytes.Equal(args.CallerAddr, l.liquidStakingSCAddress) {
+		l.eei.AddReturnMessage("invalid caller")
+		return vmcommon.UserError
+	}
+	if args.CallValue.Cmp(zero) != 0 {
+		l.eei.AddReturnMessage("not a payable function")
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/liquidStaking.go:130]: "function is not payable in eGLD" to keep the same message format used in all the other methods from this file

File: vm/systemSmartContracts/liquidStaking.go [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+	return vmcommon.Ok
+}
+
+func (l *liquidStaking) claimDelegatedPosition(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if args.CallValue.Cmp(zero) != 0 {
+		l.eei.AddReturnMessage("function is not payable in eGLD")
+		return vmcommon.UserError
+	}
+	if len(args.Arguments) == 0 {
+		l.eei.AddReturnMessage("not enough arguments")
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/liquidStaking.go:177]: "invalid number of arguments" ?

File: vm/systemSmartContracts/liquidStaking.go [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+	}
+	if len(args.ESDTTransfers) > 0 {
+		l.eei.AddReturnMessage("function is not payable in ESDT")
+		return vmcommon.UserError
+	}
+
+	return vmcommon.Ok
+}
+
+func (l *liquidStaking) claimRewardsFromDelegatedPosition(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/liquidStaking.go:188]: claimRewardsViaLiquidStaking is already implemented in delegation.go. This method shouldn't be also implemented?

File: vm/systemSmartContracts/liquidStaking.go [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+	returnCode := l.checkArgumentsWhenPositionIsInput(args)
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+	return vmcommon.Ok
+}
+
+func (l *liquidStaking) reDelegateRewardsFromPosition(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/liquidStaking.go:196]: These next methods will be implemented in this PR ?

File: vm/systemSmartContracts/proto/liquidStaking.proto [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+option go_package = "systemSmartContracts";
+option (gogoproto.stable_marshaler_all) = true;
+
+import "github.com/gogo/protobuf/gogoproto/gogo.proto";
+
+message LiquidStakingAttributes {
+  bytes  ContractAddress   = 1 [(gogoproto.jsontag) = "ContractAddress"];
+  uint32 RewardsCheckpoint = 2 [(gogoproto.jsontag) = "RewardsCheckpoint"];
+}
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/proto/liquidStaking.proto:13]: Add empty line

File: vm/address.go [hunk 9]
```diff
@@ -21,5 +21,8 @@ var EndOfEpochAddress = []byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 255,
 // DelegationManagerSCAddress is the hard-coded address for the delegation manager smart contract
 var DelegationManagerSCAddress = []byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 255, 255}
 
+// LiquidStakingSCAddress is the hard-coded address for the delegation token smart contract
+var LiquidStakingSCAddress = []byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 5, 255, 255}
```
Comment (iulianpascalau) [inline vm/address.go:25]: maybe add in the /examples/address_test.go TestSystemSCsAddressesAndSpecialAddresses this address?

File: vm/systemSmartContracts/eei.go [hunk 10]
```diff
@@ -239,7 +234,7 @@ func (host *vmContext) Transfer(
 	}
 	destAcc.OutputTransfers = append(destAcc.OutputTransfers, outputTransfer)
 
-	return nil
+	return
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/eei.go:237]: useless return

File: vm/systemSmartContracts/eei.go [hunk 11]
```diff
@@ -447,6 +436,38 @@ func (host *vmContext) AddReturnMessage(message string) {
 	host.returnMessage += "@" + message
 }
 
+// ProcessBuiltInFunction will process the given built in function and will merge the generated output accounts and logs
+func (host *vmContext) ProcessBuiltInFunction(
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/eei.go:440]: unit test?

File: vm/systemSmartContracts/delegation.go [hunk 12]
```diff
@@ -2599,6 +2615,128 @@ func (d *delegation) getMetaData(args *vmcommon.ContractCallInput) vmcommon.Retu
 	return vmcommon.Ok
 }
 
+func (d *delegation) basicCheckForLiquidStaking(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/delegation.go:2618]: missing tests

File: vm/systemSmartContracts/delegation.go [hunk 13]
```diff
@@ -2599,6 +2615,128 @@ func (d *delegation) getMetaData(args *vmcommon.ContractCallInput) vmcommon.Retu
...
+	if len(args.Arguments) != 3 {
+		d.eei.AddReturnMessage("not enough arguments")
+		return vmcommon.UserError
+	}
+
+	address := args.Arguments[0]
+	value := big.NewInt(0).SetBytes(args.Arguments[1])
+	checkPoint := uint32(big.NewInt(0).SetBytes(args.Arguments[2]).Uint64())
+
+	totalRewards, err := d.computeRewards(checkPoint, false, value)
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/delegation.go:2715]: so the claim operation via liquid staking will also take all unclaimed rewards (if existing) before creating the liquid position, right?

File: epochStart/metachain/systemSCs.go [hunk 14]
```diff
@@ -71,6 +71,7 @@ type systemSCProcessor struct {
 	esdtEnableEpoch                uint32
 	saveJailedAlwaysEnableEpoch    uint32
 	governanceEnableEpoch          uint32
+	builtInOnMetaEnableEpoch       uint32
```
Comment (iulianpascalau) [inline epochStart/metachain/systemSCs.go:74]: undefined epoch in enableEpochs.toml + print in the /node/nodeRunner.go

File: vm/systemSmartContracts/delegation.go [hunk 15]
```diff
@@ -64,6 +64,8 @@ type delegation struct {
 	validatorToDelegationEnableEpoch   uint32
 	flagReDelegateBelowMinCheck        atomic.Flag
 	reDelegateBelowMinCheckEnableEpoch uint32
+	liquidStakingEnableEpoch           uint32
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/delegation.go:67]: undefined epoch in enableEpochs.toml + print in the /node/nodeRunner.go

File: vm/systemSmartContracts/esdt.go [hunk 16]
```diff
@@ -217,6 +227,67 @@ func (e *esdt) init(_ *vmcommon.ContractCallInput) vmcommon.ReturnCode {
 	return vmcommon.Ok
 }
 
+func (e *esdt) initDelegationESDTOnMeta(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/esdt.go:230]: unit tests

File: vm/systemSmartContracts/liquidStaking.go [hunk 17]
```diff
@@ -0,0 +1,244 @@
...
+	gasCost                  vm.GasCost
+	marshalizer              marshal.Marshalizer
+	hasher                   hashing.Hasher
+	mutExecution             sync.RWMutex
+	liquidStakingEnableEpoch uint32
+	flagLiquidStaking        atomic.Flag
+}
+
+// ArgsNewLiquidStaking defines the arguments to create the liquid staking smart contract
+type ArgsNewLiquidStaking struct {
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/liquidStaking.go:37]: according to our last talk about naming convention, this should be `LiquidStakingArgs`

File: vm/systemSmartContracts/liquidStaking.go [hunk 18]
```diff
@@ -0,0 +1,244 @@
...
+	l.mutExecution.RLock()
+	defer l.mutExecution.RUnlock()
+
+	err := CheckIfNil(args)
+	if err != nil {
+		l.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+	if !l.flagLiquidStaking.IsSet() {
+		l.eei.AddReturnMessage("liquid staking contract is not enabled")
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/liquidStaking.go:101]: I think this is not backwards compatible if we end up having a transaction towards the liquid staking contract before we have this code up and running.

File: vm/systemSmartContracts/liquidStaking.go [hunk 19]
```diff
@@ -0,0 +1,244 @@
...
+	"github.com/ElrondNetwork/elrond-go-core/marshal"
+	"github.com/ElrondNetwork/elrond-go/config"
+	"github.com/ElrondNetwork/elrond-go/vm"
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+)
+
+const tokenIDKey = "tokenID"
+const noncePrefix = "n"
+
+type liquidStaking struct {
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/liquidStaking.go:22]: is this sc complete? What it does is that it just have some checks on some operations. 
\+ do not forget about tests.

File: vm/systemSmartContracts/proto/liquidStaking.proto (hunk 8 above)
Comment (iulianpascalau) [inline vm/systemSmartContracts/proto/liquidStaking.proto:13]: empty line

**Review by iulianpascalau — COMMENTED** [review]
minor stuff

File: vm/systemSmartContracts/delegation.go [hunk 20]
```diff
@@ -2599,6 +2615,128 @@ func (d *delegation) getMetaData(args *vmcommon.ContractCallInput) vmcommon.Retu
...
+	}
+	if args.CallValue.Cmp(zero) != 0 {
+		d.eei.AddReturnMessage("call value must be 0")
+		return vmcommon.UserError
+	}
+	if len(args.Arguments) < 2 {
+		d.eei.AddReturnMessage("not enough arguments")
+		return vmcommon.UserError
+	}
+	err := d.eei.UseGas(d.gasCost.MetaChainSystemSCsCost.DelegationOps)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/delegation.go:2635]: we also use liquid staking ops cost as well - from liquid staking manager contract.

File: vm/systemSmartContracts/delegation.go [hunk 21]
```diff
@@ -1816,12 +1824,30 @@ func (d *delegation) computeAndUpdateRewards(callerAddress []byte, delegator *De
 
 	isOwner := d.isOwner(callerAddress)
 
+	totalRewards, err := d.computeRewards(delegator.RewardsCheckpoint, isOwner, activeFund.Value)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/delegation.go:1827]: done

File: vm/systemSmartContracts/delegation.go [hunk 22]
```diff
@@ -2599,6 +2615,128 @@ func (d *delegation) getMetaData(args *vmcommon.ContractCallInput) vmcommon.Retu
...
+		return vmcommon.UserError
+	}
+	if len(args.Arguments) < 2 {
+		d.eei.AddReturnMessage("not enough arguments")
+		return vmcommon.UserError
+	}
+	err := d.eei.UseGas(d.gasCost.MetaChainSystemSCsCost.DelegationOps)
+	if err != nil {
+		d.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/delegation.go:2638]: done

File: vm/systemSmartContracts/delegation.go [hunk 23]
```diff
@@ -2599,6 +2615,128 @@ func (d *delegation) getMetaData(args *vmcommon.ContractCallInput) vmcommon.Retu
...
+}
+
+func (d *delegation) claimRewardsViaLiquidStaking(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	returnCode := d.basicCheckForLiquidStaking(args)
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+
+	if len(args.Arguments) != 3 {
+		d.eei.AddReturnMessage("not enough arguments")
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/delegation.go:2707]: done

File: vm/systemSmartContracts/delegation.go [hunk 24]
```diff
@@ -2599,6 +2615,128 @@ func (d *delegation) getMetaData(args *vmcommon.ContractCallInput) vmcommon.Retu
...
+	if err != nil {
+		d.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	d.eei.Transfer(args.CallerAddr, address, totalRewards, nil, 0)
+	return vmcommon.Ok
+}
+
+func (d *delegation) reDelegateRewardsViaLiquidStaking(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/delegation.go:2725]: yes.

File: vm/systemSmartContracts/esdt.go [hunk 25]
```diff
@@ -217,6 +227,67 @@ func (e *esdt) init(_ *vmcommon.ContractCallInput) vmcommon.ReturnCode {
...
+	token.SpecialRoles = append(token.SpecialRoles, esdtRole)
+
+	err = e.saveToken(tokenIdentifier, token)
+	if err != nil {
+		e.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	err = e.eei.ProcessBuiltInFunction(
+		e.eSDTSCAddress,
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:276]: done

File: vm/systemSmartContracts/liquidStaking.go (hunk 1 above)
Comment (sasurobert) [reply vm/systemSmartContracts/liquidStaking.go:81]: yes. There is no reason to enable esdt on metachain without liquid staking

File: vm/systemSmartContracts/liquidStaking.go (hunk 2 above)
Comment (sasurobert) [reply vm/systemSmartContracts/liquidStaking.go:100]: actually we have a check before - this is as plus. in system VM we have a check before reaching this

Comment (sasurobert) [reply vm/systemSmartContracts/liquidStaking.go:100]: I would leave as it is.

File: vm/systemSmartContracts/liquidStaking.go (hunk 3 above)
Comment (sasurobert) [reply vm/systemSmartContracts/liquidStaking.go:157]: done in next PR

File: vm/systemSmartContracts/liquidStaking.go (hunk 5 above)
Comment (sasurobert) [reply vm/systemSmartContracts/liquidStaking.go:177]: this has concrete check on next PR.

File: vm/systemSmartContracts/liquidStaking.go (hunk 6 above)
Comment (sasurobert) [reply vm/systemSmartContracts/liquidStaking.go:188]: methods from delegation.go can be called by liquidStaking contract only

File: vm/systemSmartContracts/liquidStaking.go (hunk 7 above)
Comment (sasurobert) [reply vm/systemSmartContracts/liquidStaking.go:196]: next PR

File: vm/systemSmartContracts/proto/liquidStaking.proto (hunk 8 above)
Comment (sasurobert) [reply vm/systemSmartContracts/proto/liquidStaking.proto:13]: done

File: vm/address.go (hunk 9 above)
Comment (sasurobert) [reply vm/address.go:25]: done

File: vm/systemSmartContracts/eei.go (hunk 10 above)
Comment (sasurobert) [reply vm/systemSmartContracts/eei.go:237]: done

File: vm/systemSmartContracts/eei.go (hunk 11 above)
Comment (sasurobert) [reply vm/systemSmartContracts/eei.go:440]: next PRs

File: vm/systemSmartContracts/delegation.go (hunk 12 above)
Comment (sasurobert) [reply vm/systemSmartContracts/delegation.go:2618]: next PRs. want to have the full picture first

File: vm/systemSmartContracts/delegation.go (hunk 13 above)
Comment (sasurobert) [reply vm/systemSmartContracts/delegation.go:2715]: yes.

File: epochStart/metachain/systemSCs.go (hunk 14 above)
Comment (sasurobert) [reply epochStart/metachain/systemSCs.go:74]: we have it: BuiltInFunctionOnMetaEnableEpoch.

Comment (sasurobert) [reply epochStart/metachain/systemSCs.go:74]: it was added some PRs before.

File: vm/systemSmartContracts/delegation.go (hunk 15 above)
Comment (sasurobert) [reply vm/systemSmartContracts/delegation.go:67]: BuiltInFunctionOnMetaEnableEpoch is used for this

File: vm/systemSmartContracts/esdt.go (hunk 16 above)
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:230]: we have in systemSCs.go - but will come in next PRs

File: vm/systemSmartContracts/liquidStaking.go (hunk 18 above)
Comment (sasurobert) [reply vm/systemSmartContracts/liquidStaking.go:101]: We do have this backward compatible - it is resolved directly in systemVM. line 108-110. It was done for delegation contract first. And contract have activation. They implement CanUseContract()

File: vm/systemSmartContracts/liquidStaking.go (hunk 19 above)
Comment (sasurobert) [reply vm/systemSmartContracts/liquidStaking.go:22]: No. this is incomplete.

File: vm/systemSmartContracts/proto/liquidStaking.proto (hunk 8 above)
Comment (sasurobert) [reply vm/systemSmartContracts/proto/liquidStaking.proto:13]: done

File: vm/systemSmartContracts/delegation.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply vm/systemSmartContracts/delegation.go:2727]: ok

File: vm/systemSmartContracts/delegation.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply vm/systemSmartContracts/delegation.go:2637]: ok

File: vm/systemSmartContracts/liquidStaking.go (hunk 1 above)
Comment (SebastianMarian) [reply vm/systemSmartContracts/liquidStaking.go:81]: ok

File: vm/systemSmartContracts/liquidStaking.go (hunk 2 above)
Comment (SebastianMarian) [reply vm/systemSmartContracts/liquidStaking.go:100]: ok

File: vm/systemSmartContracts/liquidStaking.go (hunk 3 above)
Comment (SebastianMarian) [reply vm/systemSmartContracts/liquidStaking.go:157]: ok

File: vm/systemSmartContracts/liquidStaking.go (hunk 5 above)
Comment (SebastianMarian) [reply vm/systemSmartContracts/liquidStaking.go:177]: ok

File: vm/systemSmartContracts/liquidStaking.go (hunk 6 above)
Comment (SebastianMarian) [reply vm/systemSmartContracts/liquidStaking.go:188]: ok

**Review by SebastianMarian — APPROVED** [review]

File: vm/systemSmartContracts/delegation.go (hunk 13 above)
Comment (iulianpascalau) [reply vm/systemSmartContracts/delegation.go:2717]: good

File: vm/systemSmartContracts/liquidStaking.go (hunk 18 above)
Comment (iulianpascalau) [reply vm/systemSmartContracts/liquidStaking.go:101]: ok

File: vm/systemSmartContracts/liquidStaking.go (hunk 19 above)
Comment (iulianpascalau) [reply vm/systemSmartContracts/liquidStaking.go:22]: ok

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**